File: backend/feature/post/src/application/use_case/create_label_use_case.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::{
    application::{error::post_error::PostError, gateway::label_repository::LabelRepository},
    domain::entity::label::Label,
};

#[async_trait]
pub trait CreateLabelUseCase: Send + Sync {
    async fn execute(&self, label: Label) -> Result<i32, PostError>;
}

pub struct CreateLabelUseCaseImpl {
    label_repository: Arc<dyn LabelRepository>,
}

impl CreateLabelUseCaseImpl {
    pub fn new(label_repository: Arc<dyn LabelRepository>) -> Self {
        Self { label_repository }
    }
}

#[async_trait]
impl CreateLabelUseCase for CreateLabelUseCaseImpl {
    async fn execute(&self, label: Label) -> Result<i32, PostError> {
        self.label_repository.create_label(label).await
    }
}

File: frontend/src/lib/label/application/useCase/createLabelUseCase.ts
import type {
	CreateLabelParams,
	LabelRepository,
} from '$lib/label/application/gateway/labelRepository';
import type { Label } from '$lib/label/domain/entity/label';

export class CreateLabelUseCase {
	constructor(private readonly labelRepository: LabelRepository) {}

	async execute(params: CreateLabelParams): Promise<Label> {
		return this.labelRepository.createLabel(params);
	}
}
